zipline/data/fx/base.py

```diff
@@ -63,10 +103,44 @@ def get_rate_scalar(self, rate, quote, base, dt):
         rate : np.float64
             Exchange rate from base -> quote on dt.
         """
-        rates_array = self.get_rates(
+        rates_2d = self.get_rates(
             rate,
             quote,
             bases=np.array([base], dtype=object),
             dts=pd.DatetimeIndex([dt], tz='UTC'),
         )
-        return rates_array[0, 0]
+        return rates_2d[0, 0]
+
+    @default
+    def get_rates_columnar(self, rate, quote, bases, dts):
+        """
+        Load a 1D array of FX rates.
+
+        Parameters
+        ----------
+        rate : str
+            Name of the rate to load.
+        quote : str
+            Currency code of the currency to convert into.
+        bases : np.array[object]
+            Array of codes of the currencies to convert from. The same currency
+            may appear multiple times.
+        dts : np.DatetimeIndex
+            Datetimes for which to load rates. The same value may appear
+            multiple times, but datetimes must be sorted in ascending order and
+            localized to UTC.
+        """
+        if len(bases) != len(dts):
+            raise ValueError(
+                "len(bases) ({}) != len(dts) ({})".format(len(bases), len(dts))
+            )
+
+        unique_bases, bases_ix = np.unique(bases, return_inverse=True)
+        unique_dts, dts_ix = np.unique(dts.values, return_inverse=True)
+        rates_2d = self.get_rates(
+            rate,
+            quote,
+            unique_bases,
+            pd.DatetimeIndex(unique_dts, tz='utc')
+        )
+        return rates_2d[dts_ix, bases_ix]
```

Comment:
I could be wrong about this, but it seems like in the columnar case we might load a bunch more data than we need since `get_rates` computes the cartesian product but the column is only a subset of that.  I assume this doesn't really matter, but was curious if that is a concern at all.

Reply:
> it seems like in the columnar case we might load a bunch more data than we need since get_rates computes the cartesian product but the column is only a subset of that.

Yup, this is definitely a possibility. The worst case scenario here would be something like a columnar query where every day has a different currency in it. We'd end up loading all the currencies for every day, even though we only needed one day of each currency. I'm not too worried about this because:

1. In our expected workloads, I expect that if we need a currency on day N, we'll probably need it for nearby days. In particular, most datasets will have static currencies, which means the most common case will be that if we ever need a currency we'll need it every day.
2. Even in the pathological case, it's not clear that we can do much better than what we're doing here, because we're going to want to use compression and chunking in the HDF5 files. What that means in practice is that the data arrays get broken up into regular sized pieces, and each piece gets compressed and stored separately. If you load any value out of the array, you have to decompress and load the entire chunk holding the value, so what dominates performance isn't so much the number of **values** read: it's the number of distinct **chunks** read. So, most likely, even if we tried to more precise reads here, we'd end up doing the same amount of IO, and we'd likely have higher overheads from trying to read smaller chunks of the file.